src/tls_link.c

```diff
@@ -0,0 +1,169 @@
+    }
+
+    tls_handshake_state hs_state = tls->engine->api->handshake_state(tls->engine->engine);
+    if (hs_state == TLS_HS_CONTINUE) {
+        UM_LOG(VERB, "continuing TLS handshake(%zd bytes received)", nread);
+        uv_buf_t buf;
+        buf.base = malloc(32 * 1024);
+        tls_handshake_state st =
+                tls->engine->api->handshake(tls->engine->engine, b->base, nread, buf.base, &buf.len, 32 * 1024);
+
+        UM_LOG(VERB, "continuing TLS handshake(sending %zd bytes, st = %d)", buf.len, st);
+        if (buf.len > 0) {
+            uv_link_propagate_write(l->parent, l, &buf, 1, NULL, tls_write_cb, buf.base);
+        }
+        else {
+            free(buf.base);
+        }
+
+        if (st == TLS_HS_COMPLETE) {
+            UM_LOG(VERB, "handshake completed");
+            tls->hs_cb(tls, TLS_HS_COMPLETE);
+        }
+        else if (st == TLS_HS_ERROR) {
+            char err[1024];
+            int errlen = 0;
+            if (tls->engine->api->strerror) {
+                errlen = tls->engine->api->strerror(tls->engine->engine, err, sizeof(err));
+            }
+            UM_LOG(ERR, "TLS handshake error %*.*s", errlen, errlen, err);
+            tls->hs_cb(tls, st);
+            uv_link_propagate_read_cb(l, UV_ECONNABORTED, NULL);
+        }
+    }
+    else if (hs_state == TLS_HS_COMPLETE) {
+        uv_buf_t read_buf;
+        uv_link_propagate_alloc_cb(l, 32 * 1024, &read_buf);
+
+        size_t readbuflen = read_buf.len;
+        read_buf.len = 0;
+
+        size_t out_bytes;
+        char *inptr = b->base;
+        size_t inlen = nread;
+        int rc;
+        do {
+            rc = tls->engine->api->read(tls->engine->engine, inptr, inlen,
+                    read_buf.base + read_buf.len, &out_bytes, readbuflen - read_buf.len);
+
+            UM_LOG(VERB, "produced %zd application byte (rc=%d)", out_bytes, rc);
+            read_buf.len += out_bytes;
+            inptr = NULL;
+            inlen = 0;
+        } while (rc == TLS_MORE_AVAILABLE && out_bytes > 0);
+
+        uv_link_propagate_read_cb(l, read_buf.len, &read_buf);
+    }
+    else {
+        UM_LOG(VERB, "hs_state = %d", hs_state);
+    }
+
+    if (b != NULL && b->base != NULL) {
+        free(b->base);
+    }
+}
+
+static int tls_write(uv_link_t *l, uv_link_t *source, const uv_buf_t bufs[],
+                     unsigned int nbufs, uv_stream_t *send_handle, uv_link_write_cb cb, void *arg) {
+    tls_link_t *tls = (tls_link_t *) l;
+    uv_buf_t buf;
+    buf.base = malloc(32 * 1024);
+    tls->engine->api->write(tls->engine->engine, bufs[0].base, bufs[0].len, buf.base, &buf.len, 32 * 1024);
+    int rc = uv_link_propagate_write(l->parent, l, &buf, 1, NULL, tls_write_cb, buf.base);
+
+    cb(source, 0, arg);
+
+    return rc;
+}
+
+static void tls_close(uv_link_t *l, uv_link_t *source, uv_link_close_cb close_cb) {
+    UM_LOG(VERB, "closing TLS link");
+//    um_http_t *clt = l->data;
+//
+//    clt->tls->api->free_engine(clt->engine);
```

Comment:
? not needed or is there a follow-up needed?

Reply:
not needed client is in charge of the engine lifecycle